### DISCUSSION TITLE
PNP (People) Search - Is it possible to exclude a property from being searched?

### DISCUSSION THREAD

**Discussion by StephenHoughton4** [microsoft-search/pnp-modern-search#4004] [Q&A] [answered]
### Background
We have had a solution in place since mid 2023, which has been brought to our attention does not function as intended, especially when there are multiple search terms.

Originally we had results that appeared that should not have done when searching one word (https://github.com/microsoft-search/pnp-modern-search/discussions/2922)

It was determined that the issue was the Memberships property, which seems to display all AD Groups a person is part of. We then included _**(({SearchTerms}*) AND NOT Memberships:({SearchTerms}*))**_. 

This seemed to fix the issue, for single word searches and for a name and another property (office or department). The current issue is where we search for London Finance, where we get missing results and extra results.

### Technical
Breaking down the query we have 3 users and I will explain the situation

User1
Department: Finance | Office: London | Memberships: LondonUsers
Status: Showing as expected. 
Technical: Memberships does not match both terms, so ({SearchTerms}*) is used for the search.

User2
Department: Finance | Office: Madrid | Memberships: LondonUsers
Status: Showing incorrectly. 
Technical: Memberships does not match both terms, so ({SearchTerms}*) is used for the search, which includes searching Memberships.

User3
Department: Finance | Office: London | Memberships: LondonUsers; FinanceTeam
Status: Not Showing, but should be. 
Technical: Memberships matches both terms, so is excluded from results.

### The Question
Is it possible to exclude the property Memberships from being searched?
Can we search the department and location together, example (Department+location:({SearchTerms}*)  ),

I realise no one has asked this question, so there may be something fundamental I have missed, so any help will be appreciated.

**Comment by kasperbolarsen** [comment]
Department is searchable and refinable out of the box, and so is Office location (BaseOfficeLocation). Both are usually used as a refiner.
You can combine a search, like Department:{SearchTerms}* OR BaseOfficeLocation:{SearchTerms}*, but that means that you exclude ALL the other fields, which is rarely a good idea.
I would recommend working with the sequence the data is presented in instead, making the most relevant hits show out in top. Search this repo and/or the web for XRANK and you should be good :-) 
![image](https://github.com/user-attachments/assets/c0359a2b-055e-4391-8ce4-e7821b12f384)

**Reply by StephenHoughton4** [reply]
Hi,

Our current Query Text is :
```
((({SearchTerms}*) AND NOT Memberships:({SearchTerms}*))
OR UserName:({searchTerms}*)
OR JobTitle:({searchTerms}*)
OR AccountName:({searchTerms}*)
OR Department:({searchTerms}*)
OR Office:({searchTerms}*)
OR Title:({searchTerms}*)
OR PersonInitials:({searchTerms}*)
OR FirstName:({searchTerms}*)
OR LastName:({searchTerms}*)
OR Location:({SearchTerms}*)
OR (MobilePhone:{SearchTerms}*)
)

-accountname:adm* -username:svc* -Title:Test* Workemail:ourdomain.com LastName:* -"sps-hidefromaddresslists":1
JobTitle:* -JobTitle:"Test" -JobTitle:"Shared Mailbox"

(-"sps-usertype":1 AND -"sps-hidefromaddresslists":1 AND -"sps-recipienttypedetails"=4 AND -"sps-recipienttypedetails"=16 AND -"sps-recipienttypedetails"=32 AND -"sps-recipienttypedetails"=8589934592 AND -"sps-recipienttypedetails"=17179869184 AND -"sps-recipienttypedetails"=2199023255552)
```

This is using the function of OR as you suggested, but London Finance does not seem to hit these properties properly. As the Department is just one of the terms, and Office is one of the other terms.

I was assuming the AD Memberships could be excluded somehow, we attempted the XRANK option previously, but the results are in a wild order, As in will show London office first, but there is no order on the Names. It would be good if each result had a rank of 1, then this was increased by Office being London, so you could organise by the LastName also, but as each result has a Ranked number (e.g. 1.0002013), this means the results cannot be ordered any better.

The issue with removing the NOT Memberships option, is that the results when searching for London, would include anyone in a Membership group (e.g. LondonHRPolicies for a Global HR Person based elsewhere).

**Reply by StephenHoughton4** [reply]
London Finance is not hitting correctly. I have figured out why this is, but not how to fix the issue.
({searchTerms}*) converts into London and Finance*
So Memberships Groups such as LondonUsers would not actually be hit, But FinanceUsers and "London-Admin" would.

My initial line of code:
`(({SearchTerms}*) AND NOT Memberships:({SearchTerms}*))`
would mean some users in 2 memebrships groups will not be displayed, even if other properties match, examples:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/YSH/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/YSH/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">


Person | Office | Department | Memberships | Shows in Search
-- | -- | -- | -- | --
Amy | London | Finance | LondonUsers;     FinanceUsers | YES
Bob | London | Finance | London-Admin;     FinanceUsers | NO
Carl | Paris | Finance | ParisUsers;      FinanceUsers | NO
Dani | Paris | Finance | ParisUsers;      London-Auditors | YES
Earl | London | Lawyer | LondonUsers;     FinanceLaw | YES
Fred | Paris | Lawyer | ParisUsers;     FinanceLaw | NO



</body>

</html>

Reviewing the above (when Searching **London Finance**)
Amy, Dani and Earl show in the Search results. It is expected to see Amy and Bob only in the results.

**_Amy_** - Office and Finance hit the Search Parameters, the exclusion of Memberships is not activated as only one term is matched "FinanceUsers")
**_Bob_** - Office and Finance hit the Search Parameters, the exclusion of Memberships is activated as both terms are in the memberships ("London-..." and "FInance..."). The rest of the query then searches individual properties for both terms and will not find them together (unless JobTitle was "London Finance Controller" then this person would show again.
**_Carl_** - Has no information about London, so does not show. If school was listed as "London College" then they would appear.
**_Dani_** - exclusion based on Memberships is not activated as does not contain both terms, so ({SearchTerms}*) is activated, and matches the Department for "Finance" and the Memberships group called "London-..."
**_Earl_** - similar to Dani fails the Exclusion, but then matches 2 terms one in Office and one in Memberships. So is displayed incorrectly.
**_Fred_** - As no mention of London, the only term hit is in Memberships, so will not satisfy the search Criteria.

My question was how do we get only Amy and Bob to show in the results, and not show the other 2 members that show incorrectly.

**Comment by kasperbolarsen** [comment]
In this case it sounds like the best approach is either to exclude the Memberof property from the index or just specify each property you want to query:

UserName:({searchTerms}*)
OR JobTitle:({searchTerms}*)
OR AccountName:({searchTerms}*)
OR Department:({searchTerms}*)
OR Office:({searchTerms}*)
OR Title:({searchTerms}*)
OR PersonInitials:({searchTerms}*)
OR FirstName:({searchTerms}*)
OR LastName:({searchTerms}*)
OR Location:({SearchTerms}*)
OR (MobilePhone:{SearchTerms}*)
)

-accountname:adm* -username:svc* -Title:Test* Workemail:ourdomain.com LastName:* -"sps-hidefromaddresslists":1
JobTitle:* -JobTitle:"Test" -JobTitle:"Shared Mailbox"

**Reply by StephenHoughton4** [reply]
Thanks Kasper,

I tried similar and if you search London Finance, expecting to find All people in Finance based in London, they would not appear, as office does not have both London and Finance within. So unfortunately this wont work.

I did see in the admin of Sharepoint the Managed Properties, Memberships is listed as Searchable, can this be changed? (It may be its locked for my account to do so, but I can get someone to amend if its possible)
![image](https://github.com/user-attachments/assets/5a3c14a9-dd2b-4b9a-b4c9-addf83f62eb3)
Admin Console > More Features > Search > Manage Search Schema

**Comment by StephenHoughton4** [comment, marked as answer]
To answer my question, it is not possible to exclude a property from being searched.

However, there is a workaround solution for this. Create a new Managed Property that is Searchable, this will take time to replicate through to all the users, so will not be an instant fix.

### How to Create a Managed Property
This is quite Simple, which an Office 365 Admin should be able to perform.
Head into Admin Console > More Features > Search > Manage Search Schema
Then click on New Managed Property

For this example I will Use the Name **_PNPPeopleCriteria_**
Add the name, a short description (so it doesn't get deleted by any IT Clean-up exercise) and then make the property Searchable.
![image](https://github.com/user-attachments/assets/6732cb96-5504-4c66-8315-5359119bb1b2)
The next part is to Set the Crawled Properties to "include content from all crawled Properties".
Next Add a mapping. you could add all from the "People" category, or look which Managed Properties you want and which crawled properties feed into them.

### Using the new Property
This will not be instantaneous, as will take some time for the new property to propagate through.
So what you would want to do is perform a search of _PNPPeopleCriteria:*_  this will return all People that have this property.
Compare this to the expected list of results available (example searching LastName:*). When these match you can then move to updating the People Search page.

Edit the Query Template to only reference the new property (Keeping any exclusion rules you have in place):
```
PNPPeopleCriteria:({SearchTerms}*)

-"Below are the Exclusion and inclusion rules, which I left un-amended":*

-accountname:admin* -username:svc* -Title:Test* Workemail:ourdomain.com LastName:* -"sps-hidefromaddresslists":1
JobTitle:* -JobTitle:"Test" -JobTitle:"Shared Mailbox"

(-"sps-usertype":1 AND -"sps-hidefromaddresslists":1 AND -"sps-recipienttypedetails"=4 AND -"sps-recipienttypedetails"=16 AND -"sps-recipienttypedetails"=32 AND -"sps-recipienttypedetails"=8589934592 AND -"sps-recipienttypedetails"=17179869184 AND -"sps-recipienttypedetails"=2199023255552)
```

If you saw my previous code, I have added the PNPPeopleCriteria:({SearchTerms}*) in place of the below (which had the erroneous behaviour):
```
((({SearchTerms}*) AND NOT Memberships:({SearchTerms}*)) 
OR UserName:({searchTerms}*) OR JobTitle:({searchTerms}*) OR AccountName:({searchTerms}*) OR Department:({searchTerms}*) 
OR Office:({searchTerms}*) OR Title:({searchTerms}*) OR PersonInitials:({searchTerms}*) OR FirstName:({searchTerms}*) 
OR LastName:({searchTerms}*) OR Location:({SearchTerms}*) OR MobilePhone:({SearchTerms}*) 
)
```

Hope this helps anyone else facing this error. Also it is worth noting that the * Applies only to the last Term entered not both, so this may need reviewing if you do not intend this behaviour.

**Reply by wobba** [reply]
You should do a docs PR for this as a scenario :)

**Reply by StephenHoughton4** [reply]
I will do, off to Google how to do one :P
